File: packages/dart/lib/src/catalog/offering_catalog.dart
/// Host-declared product / offering / add-on codes this binary knows how to gate.
///
/// At runtime, [LicenseEntitlements] intersects this set with the verified license
/// JWT. Unknown JWT codes are ignored; catalog codes missing from the JWT fail closed.
/// Omit the catalog to keep JWT-only gating.
class OfferingCatalog {
  const OfferingCatalog({
    required this.productIds,
    required this.offeringCodes,
    required this.addonCodes,
  });

  final Set<String> productIds;
  final Set<String> offeringCodes;
  final Set<String> addonCodes;

  /// True when [productId] is in this catalog.
  bool knowsProduct(String productId) => productIds.contains(productId);

  /// True when [offeringCode] is in this catalog.
  bool knowsOffering(String offeringCode) =>
      offeringCodes.contains(offeringCode.trim());

  /// True when [addonCode] matches a catalog add-on (case-insensitive).
  bool knowsAddon(String addonCode) {
    final needle = addonCode.trim().toLowerCase();
    return addonCodes.any((a) => a.toLowerCase() == needle);
  }
}

File: packages/javascript/packages/browser-sdk/src/billing/offering-catalog.ts
/**
 * Host-declared product / offering / add-on codes this binary knows how to gate.
 *
 * Runtime gates are `static catalog ∩ verified license JWT`.
 * Unknown JWT codes are ignored; catalog codes missing from the JWT fail closed.
 */
export type OfferingCatalog = {
  productIds: readonly string[];
  offeringCodes: readonly string[];
  addonCodes: readonly string[];
};

function inSet(haystack: readonly string[], needle: string, caseInsensitive = false): boolean {
  const n = needle.trim();
  if (!caseInsensitive) return haystack.includes(n);
  const lower = n.toLowerCase();
  return haystack.some((h) => h.toLowerCase() === lower);
}

/** True when [catalog] includes [productId]. */
export function catalogKnowsProduct(catalog: OfferingCatalog, productId: string): boolean {
  return inSet(catalog.productIds, productId);
}

/** True when [catalog] includes [offeringCode]. */
export function catalogKnowsOffering(catalog: OfferingCatalog, offeringCode: string): boolean {
  return inSet(catalog.offeringCodes, offeringCode);
}

/** True when [catalog] includes [addonCode] (case-insensitive). */
export function catalogKnowsAddon(catalog: OfferingCatalog, addonCode: string): boolean {
  return inSet(catalog.addonCodes, addonCode, true);
}
